L7LB: Extract Envoy related logic and dependencies from ServiceManager

mhofstetter opened the pull request:
Goal of this PR is to extract the Envoy related logic (and the necessary dependencies) from the `ServiceManager` and reduce the number of dependencies to Envoy's xDS server from non-envoy packages (follow up of https://github.com/cilium/cilium/pull/29892).

This has also been discussed in the PR that introduced Envoys "frontendPort filter logic" to the ServiceManager: https://github.com/cilium/cilium/pull/19647

* https://github.com/cilium/cilium/pull/19647#discussion_r886919428
* https://github.com/cilium/cilium/pull/19647#discussion_r883478969

The chosen approach is that L7 LB proxies (e.g. Envoy integration via CiliumEnvoyConfig watcher) should be able to register themselves to the `ServiceManager` to be notified (via callback function) when backends of an interested Service needs to be synced to the respective L7 loadbalancer.

In addition, the two aspects of any L7 LB integration (forwarding to proxy port & registration for backend sync) are splitted (& renamed) into separate functions (`RegisterL7LBServiceRedirect` & `RegisterL7LBServiceBackendSync` (and their deregistering counterparts)) - in order to understand this better.

This way, the Envoy backend sync logic (via xDS) and its required dependency to the xDS server, can be removed from the ServiceManager. It's up to the newly introduced `EnvoyServiceBackendSyncer` (in package `pkg/envoy`) to handle this.

Please see the individual commits for further information.

Notes:
* The registration is per service. Changing this to a global registration via Hive to be notified to all Services might be something for a follow up PR.
* Probably it's about time to split up the K8s watcher and extract the CiliumEnvoyConfig watcher. But again - something for a follow up PR.

mhofstetter committed:
service: remove parameter frontendPorts from RegisterL7LBService

Currently the parameter `frontendPorts` in the function
`ServiceManager.RegisterL7LBService` is not used from external callees
as it's sole purpose is to register a proxy port for a given service.

It's only used within the `ServiceManager` in case of calling the function
`ServiceManager.RegisterL7LBServiceBackendSync`.

Therefore, this commit removes the parameter `frontendPorts` from
`RegisterL7LBService` by properly implementing the logic of
`RegisterL7LBServiceBackendSync` instead of delegating to
`RegisterL7LBService`.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
service: introduce struct L7LBResourceName

Currently, the struct `loadbalancer.ServiceName` is used when
defining a reference to a `CiliumEnvoyConfig` during L7LB service
registration.

This commit introduces a dedicated struct `L7LBResourceName` to
prevent confusion. In addition, the corresponding fields in the
`L7LBInfo` are renamed to be Envoy agnostic.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
service: remove duplicated import

This commit removes the duplicated import `pkg/datapath/types` from
`pkg/service/service.go`.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
service: introduce callback for L7LB backend sync

Currently, `Service.RegisterL7LBServiceBackendSync` is built quite
Envoy specific, hence it also contains logic how to filter Service
backends by the frontendPorts that are passed to the method.

To be able to remove the dependency from the ServiceManager to Envoy
specific details (implementation and the dependency to the
Envoy xDS server), this commit refactors the backend sync registration
to receive a callback that gets called whenever a Service (and its backends)
changed.

This way, the Envoy specific details can be removed from the ServiceManager.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
service: remove envoyXdsServer from service manager

The service manager no longer depends on the envoy xDS server.

Hence let's remove the field from the manager and its hive cell.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
envoy: move UpsertEnvoyEndpoints logic from xDSserver to CEC watcher

Currently, the xDSServer provides the possibility to update Envoy
endpoints based on loadbalancer backend information. This logic
is only used by the CiliumEnvoyConfig watcher to update the backend
services of a CEC accordingly.

With the introduction of a service backend sync callback, this logic
can be moved from the xDS server to the CEC watcher. This way we can
get rid of unwanted dependencies from the envoy module to the loadbalancer
module.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
envoy: move EnvoyServiceBackendSyncer from watchers to envoy package

After removing the dependency from the service package (service manager)
to envoy, it's possible to move the EnvoyServiceBackendSyncer from the
k8s watchers package to the envoy package.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
envoy: provide EnvoyServiceBackendSyncer via envoy Hive Cell

This commit moves the initialization of the EnvoyServiceBackendSyncer
into the corresponding Envoy Hive Cell.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
envoy: fix error in k8s watcher

Change error message to start with lowercase.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

jrajahalme commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -160,9 +160,10 @@ type svcManager interface {
-	RegisterL7LBService(serviceName loadbalancer.ServiceName, resourceName service.L7LBResourceName, proxyPort uint16) error
-	RegisterL7LBServiceBackendSync(serviceName loadbalancer.ServiceName, resourceName service.L7LBResourceName, ports []string) error
-	RemoveL7LBService(serviceName loadbalancer.ServiceName, resourceName service.L7LBResourceName) error
+	RegisterL7LBServiceForward(serviceName loadbalancer.ServiceName, resourceName service.L7LBResourceName, proxyPort uint16) error
```
Small nit on naming, for the sake of consistency I'd c all this `RegisterL7LBServiceRedirect` instead.

mhofstetter replied:
agree: renamed all occurrences from `forward` to `redirect`

jrajahalme commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -160,9 +160,10 @@ type svcManager interface {
-	RegisterL7LBServiceBackendSync(serviceName loadbalancer.ServiceName, resourceName service.L7LBResourceName, ports []string) error
-	RemoveL7LBService(serviceName loadbalancer.ServiceName, resourceName service.L7LBResourceName) error
+	RegisterL7LBServiceForward(serviceName loadbalancer.ServiceName, resourceName service.L7LBResourceName, proxyPort uint16) error
+	RemoveL7LBServiceForward(serviceName loadbalancer.ServiceName, resourceName service.L7LBResourceName) error
```
Same here, but additionally maybe rename these functions to `Add...`/`Remove...` or `Register...`/`Deregister...`. IMO the former conventions would be more inline with the rest of the code base, rather than the latter.

mhofstetter replied:
Thanks - good point! i went with `register` / `deregister` (for forward & backend sync) as i think it's more suitable for this use case. L7 Proxy (e.g. Envoy) registers a forward or backendsync to the LB ServiceManager.

I hope this is OK for you?

jrajahalme reviewed (approved):
Small nits only, but then one structural question/suggestion: Did you consider making `envoy.XDSServer` implement the `envoy.EnvoyServiceBackendSyncer` interface? The daemon code would be a bit simpler with that?

tommyp1ckles commented on pkg/k8s/watchers/envoy_l7lb_backend_syncer.go:
```diff
@@ -0,0 +1,176 @@
+	l7lbSvcs      map[loadbalancer.ServiceName]*backendSyncInfo
+}
+
+var _ service.BackendSyncer = &EnvoyServiceBackendSyncer{}
```
nit: this could be (not sure if it actually makes any difference in compiled code).
```go
var _ service.BackendSyncer = (*EnvoyServiceBackendSyncer)(nil)
```

mhofstetter replied:
i'd like to keep this as is if possible. IMO it's more readable and should free up the memory anyway (blank identifier). (if not optimized by the compiler)

tommyp1ckles commented on pkg/envoy/l7lb_backend_syncer.go:
```diff
@@ -0,0 +1,247 @@
+		// check for port number
+		if port == strconv.Itoa(int(svc.Frontend.Port)) {
+			return map[string][]*loadbalancer.Backend{
+				port: filterPreferredBackends(svc.Backends),
```
we could avoid calling `filterPreferredBackends` twice in this instance by precomputing it.

mhofstetter replied:
This code was migrated as is. But i agree that this can be optimize => refactored

tommyp1ckles commented on pkg/envoy/l7lb_backend_syncer.go:
```diff
@@ -0,0 +1,247 @@
+}
+
+func (r *EnvoyServiceBackendSyncer) Sync(svc *loadbalancer.SVC) error {
+	r.l7lbSvcsMutex.RLock()
```
Is the mutex lock just for reading from `r.L7lbSvcs`? If so, we could probably release that prior to calling out to xDS as that can result in this holding the mutex much longer.

mhofstetter replied:
thanks, good point! 

-> refactored to free unlock mutex after reading from `l7lbsvcs`.

tommyp1ckles reviewed (approved):
Changes make sense to me, just a few random thoughts

mhofstetter committed:
envoy: precompute preferred backends in backendsync

Currently, filtering the service backends, that should be synced
to Envoy for L7 loadbalancing, might call out to `filterPreferredBackends`
multiple times - even though it would be possible to precompute them.

Therefore, this commit refactors the EnvoyL7LBBackendSyncer to
precompute the preferrred backends.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
